Implimented arXivId Parsing for PDF with arXivId

ar-rana opened the pull request:
Used the 'parse' method in ArXivIdentifier to get arXivId and added a testcase for the same using the link give in the issue.
Refs #12000 

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

ar-rana commented:
@koppor could you please review this PR and suggest changes where I am wronge.

Siedlerchr commented:
Thanks for your contribution. Please fix the check style issues first

ar-rana commented:
Hello maintainers, could you please review this PR, I have fixed the previous issues.

Siedlerchr commented:
Codewise looks good to me. You have accidentally commited the csl styles submodules, see https://devdocs.jabref.org/code-howtos/faq.html#the-problem for a solution

InAnYan commented:
Hmm, I tried to import these papers:

https://arxiv.org/abs/2406.14319
https://arxiv.org/abs/2412.06769

But it didn't import it properly, and `arXiv` was `null` all the time

ar-rana commented:
I have changed the ```getArXivId``` method to this in my local branch:
```
private String getArXivId(String arXivId) {
        if (arXivId == null) {
            arXivId = ArXivIdentifier.parse(curString).map(ArXivIdentifier::asString).orElse(null);
            if (arXivId != null) {
                if (curString.length() > arXivId.length() + 7) {
                    curString = curString.substring(arXivId.length() + 7);
                    extractYear();
                }
                return arXivId;
            }
        }
        return arXivId;
    }
```
and modified the ```ArXivIdentifier.parse``` method a little  by altering the identifier to this ```String identifier = value.split(" ")[0];``` at line 41.

this fixes the null ```arXiv``` problem that @InAnYan was encontering when externally importing the papers, but for some reason it still does not import the titles correctly except for the paper that was mentioned in the issue. 

<img width="567" alt="Screenshot 2024-12-27 195014" src="https://github.com/user-attachments/assets/c610efee-ec73-489a-8665-a479f2f6ab0a" />

InAnYan commented:
Hi, ar-rana! Thanks for working on this PR!

As holidays come soon, maintainers could be too busy these weeks, so don't worry if we give feedback too late

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -609,11 +609,15 @@ private String getDoi(String doi) {
     private String getArXivId(String arXivId) {
         if (arXivId == null) {
-            arXivId = ArXivIdentifier.parse(curString).map(ArXivIdentifier::asString).orElse(null);
+            String arXiv = curString.split(" ")[0];
```
This could lead to a null pointer if there is no whitespace and you access the index

ar-rana replied:
Hello Siedlerchr, I tested this method with empty/non-empty strings with/without whitespaces and it would only give a null pointer if the `curString` is null which does not seem to be the case here.

HoussemNasri commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -592,6 +607,24 @@ private String getDoi(String doi) {
+            String arXiv = curString.split(" ")[0];
+            arXivId = ArXivIdentifier.parse(arXiv).map(ArXivIdentifier::asString).orElse(null);
+            if (arXivId != null) {
+                if (curString.length() > arXivId.length() + 7) {
```
What does 7 stand for here? If possible, define a constant with a good name for the value 7 in this context.

Also, consider reducing the level of nesting in the code by returning early.

ar-rana replied:
7 is for the `'arxiv:'` prefix in the arxiv string, 

> If possible, define a constant with a good name for the value 7 in this context.
> 
> Also, consider reducing the level of nesting in the code by returning early.

sure, I will work on that.

HoussemNasri replied:
In that case, `ARXIV_PREFIX_LENGTH = "arxiv:".length()` would be a good name.

HoussemNasri commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -608,18 +608,16 @@ private String getDoi(String doi) {
     }
 
     private String getArXivId(String arXivId) {
+        final int ARXIV_PREFIX_LENGTH = "arxiv:".length();
```
Let's define it as static and move it outside the method to avoid reallocating the same string object everytime the method is called.

`private static final ...`

ar-rana replied:
should i move it to the `ArXivIdentifier` class instead, i think it might be more suited there

HoussemNasri replied:
In general (and in OOP more specifically) we want to keep the data as close as possible to the code that uses it, if the constant is used there then okay otherwise let's keep it in the class where it is used.

ar-rana replied:
sure, thank you

HoussemNasri commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -608,18 +608,16 @@ private String getDoi(String doi) {
-                    proceedToNextNonEmptyLine();
-                }
-                return arXivId;
+            if (arXivId != null && curString.length() > arXivId.length() + ARXIV_PREFIX_LENGTH) {
```
This is not what I meant by reduced nesting, Please read this: https://szymonkrajewski.pl/why-should-you-return-early/

ar-rana replied:
```
private String getArXivId(String arXivId) {
        if (arXivId != null) {
            return arXivId;
        }

        String arXiv = curString.split(" ")[0];
        arXivId = ArXivIdentifier.parse(arXiv).map(ArXivIdentifier::asString).orElse(null);

        if (arXivId == null || curString.length() < arXivId.length() + ARXIV_PREFIX_LENGTH) {
            return arXivId;
        }
        // The arxiv string also contains the year
        curString = curString.substring(arXivId.length() + ARXIV_PREFIX_LENGTH);
        extractYear();
        curString = "";
        proceedToNextNonEmptyLine();

        return arXivId;
    }

```


This is what I wrote based on my understanding of the article, please share you feedback

HoussemNasri commented on src/test/java/org/jabref/logic/importer/fileformat/PdfContentImporterTest.java:
```diff
@@ -129,6 +129,42 @@ British Journal of Nutrition (2008), 99, 1–11 doi: 10.1017/S0007114507795296
+                .withField((StandardField.KEYWORDS), "Test Automation Artificial Intelligence AI-assisted Test Automation Grey Literature Automated Test Generation Self-Healing Test Scripts");
+
+        String firstPageContent = """
+                arXiv:2408.06224v1 [cs.SE] 12 Aug 2024
```
Is it guaranteed that the arXiv id will be on the first line? If not we need to add at least two more tests.

- ArXiv id in a middle line (the line begins with the id)
- ArXiv ID in a middle line (the ID is somewhere in the middle of that line)

You can use JUnit 5 Parameterized Tests to reduce verbosity.

ar-rana replied:
hello houssem, I checked multiple arXiv papers(10) and all of them had their arxiv string at the last line, here you can see it at the top(sorry for the oversight, will fix that) but moving it at the last also passes the test, and since all these papers have the same format i think the arxiv string would mostly be at the end

HoussemNasri commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -364,12 +367,14 @@ Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineS
         EntryType type = StandardEntryType.InProceedings;
         if (curString.length() > 4) {
             // special case: possibly conference as first line on the page
+            arXivId = getArXivId(null);
```
This is an example why comments are to be avoided as much as possible, this line compiles fine, runs fine but the comment above it is misguiding. Please move comment to its correct location.

InAnYan commented:
Great! I've imported some arXiv paper that I had on my disk, and it had successfully extracted the ID into `eprint`!

If you want to continue working on this feature, you could try to improve it further:

Ideally, I, as a user of JabRef, would like to import a paper in PDF, and JabRef should correctly determine all the necessary metadata. Your PR works on PDF importers level: you found arXiv ID in PDF content and included in `eprint` field. This is great! And it's the first important step of improving PDF import.

Unfortunately, PDF import is not ideal (and it's a hard task on its own). Now, it would be cool if after finding the `eprint`, another fetcher would be called that would extract all metadata from arXiv service based on that ID

InAnYan commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -531,6 +538,10 @@ Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineS
             }
         }
 
+        if (arXivId != null && arXivId.contains(year)) {
```
Is there a special reason, why `year` should be nulled?

InAnYan commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -557,19 +561,25 @@ Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineS
+        if (year != null && !"0000".equals(year)) {
+            entry.setField(StandardField.YEAR, year);
+        } else if (arXivId != null) {
+            year = "20" + arXivId.substring(0, 2);
```
Still, this if clause should be deleted, as if I import the paper I mentioned before, it would probably set year to 2095, which is not right

ar-rana replied:
sure, will do that

InAnYan reviewed (changes requested):
Great! Just some small changes. The issue with `abbrev` and `csl-styles` should be resolved too

ar-rana commented:
> Great! I've imported some arXiv paper that I had on my disk, and it had successfully extracted the ID into `eprint`!
> 
> If you want to continue working on this feature, you could try to improve it further:
> 
> Ideally, I, as a user of JabRef, would like to import a paper in PDF, and JabRef should correctly determine all the necessary metadata. Your PR works on PDF importers level: you found arXiv ID in PDF content and included in `eprint` field. This is great! And it's the first important step of improving PDF import.
> 
> Unfortunately, PDF import is not ideal (and it's a hard task on its own). Now, it would be cool if after finding the `eprint`, another fetcher would be called that would extract all metadata from arXiv service based on that ID

sounds interesting, would love work on this

ar-rana commented:
@InAnYan I was working on the Arxiv Metadata method and this is what I have written so far, 
```
    private Optional<BibEntry> getMetadataFromArxiv(String arXivId) {
        PreferencesFactory preferencesFactory = new PreferencesFactory();
        ArXivFetcher arXivFetcher = new ArXivFetcher(preferencesFactory.provide().getImportFormatPreferences());

        Optional<BibEntry> bibEntry = null;
        try {
            bibEntry = arXivFetcher.performSearchById(arXivId);
        } catch (FetcherException exception) {
            ParserResult.fromError(exception);
        }

        return bibEntry;
    }
```

this method returns an `Optional<BibEntry>` using `ArXivFetcher` which can directly be returned by `getEntryFromPDFContent` method, I just wanted to know if I should proceed with this implimentation because other ways dont return all the metadata from `arXivId` but just a few fields. (will also have to significantly modify the testcase here)

Siedlerchr commented:
I fixed the submodule issues for you. Please address the comments so we can get this merged

InAnYan commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -565,13 +572,16 @@ Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineS
             entry.setField(StandardField.VOLUME, volume);
         }
-        if (number != null) {
+        if (number != null && number.chars().allMatch(Character::isDigit)) {
```
```suggestion
        if (number != null) {
```

Maybe that was a fix for some other issue?

InAnYan commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -372,11 +375,13 @@ Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineS
         String publisher = null;
 
         EntryType type = StandardEntryType.InProceedings;
         if (curString.length() > 4) {
```
`curString.length() > 4` - why bigger than 4 :sweat_smile: ?

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -557,19 +561,25 @@ Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineS
             entry.setField(StandardField.DOI, doi);
         }
+        if (arXivId != null) {
+            entry.setField(StandardField.EPRINT, arXivId);
```
Internal follow-up: https://github.com/JabRef/jabref-issue-melting-pot/issues/751

For now, we do `eprint` and (new) `eprinttype`.

koppor commented:
The issue was not fixed - we added the functionality in [`b3d8556` (#12335)](https://github.com/JabRef/jabref/pull/12335/commits/b3d85561609615da9f26a8b71f3f799c7d913090). 

We don't have the time to try it out. We "blindly" merge.

Users need to test - and a test should be added (in a follow-up PR).

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -372,11 +375,13 @@ Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineS
 
         EntryType type = StandardEntryType.InProceedings;
         if (curString.length() > 4) {
+            arXivId = getArXivId(null);
```
Passing null is really bad style. This neesd to be fixed eventually!!

I will approve nevertheless. Too much time spend on this.

koppor reviewed (approved):
See comments inside.
